extensions/SearchAPI/lib/Util.pm

```diff
@@ -0,0 +1,320 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+# This Source Code Form is "Incompatible With Secondary Licenses", as
+# defined by the Mozilla Public License, v. 2.0.
+
+package Bugzilla::Extension::SearchAPI::Util;
+
+use 5.10.1;
+use strict;
+use warnings;
+
+use base qw(Exporter);
+
+our @EXPORT_OK = qw(
+  bug_to_hash
+  comment_to_hash
+  flag_to_hash
+  named_params
+  user_to_hash
+);
+
+use constant NAMED_PREFIX => ':';
+use constant NAMED_SUFFIX => ':';
+
+# Regex borrowed from SQL::Tokenizer
+# https://metacpan.org/pod/SQL::Tokenizer
+my $tokenize_re = qr{
+    (
+        (?:--|\#)[\ \t\S]*      # single line comments
+        |
+        (?:<>|<=>|>=|<=|==|=|!=|!|<<|>>|<|>|\|\||\||&&|&|-|\+|\*(?!/)|/(?!\*)|\%|~|\^|\?)
+                                # operators and tests
+        |
+        [\[\]\(\),;.]            # punctuation (parenthesis, comma)
+        |
+        \'\'(?!\')              # empty single quoted string
+        |
+        \"\"(?!\"")             # empty double quoted string
+        |
+        "(?>(?:(?>[^"\\]+)|""|\\.)*)+"
+                                # anything inside double quotes, ungreedy
+        |
+        `(?>(?:(?>[^`\\]+)|``|\\.)*)+`
+                                # anything inside backticks quotes, ungreedy
+        |
+        '(?>(?:(?>[^'\\]+)|''|\\.)*)+'
+                                # anything inside single quotes, ungreedy.
+        |
+        /\*[\ \t\r\n\S]*?\*/      # C style comments
+        |
+        (?:[\w:@]+(?:\.(?:\w+|\*)?)*)
+                                # words, standard named placeholders, db.table.*, db.*
+        |
+        (?: \$_\$ | \$\d+ | \${1,2} )
+                                # dollar expressions - eg $_$ $3 $$
+        |
+        \n                      # newline
+        |
+        [\t\ ]+                 # any kind of white spaces
+    )
+}smx;
+
+# Some workflow borrowed from
+# https://metacpan.org/pod/DBIx::Placeholder::Named
+sub named_params {
```

Comment:
Can you import this from the place it was borrowed from instead of maintaining it ourselves? Same with `$tokenizer_re`. I think we should try to avoid maintaining this code if we can.

Reply:
I was trying to avoid bring in any more dependencies that needed and this was something that I only needed small bits of the larger library. Therefore I try to make sure I give credit the original author. We do this in several other places in the code base already. Notably Bugzilla::S3 which is a copied version of an abandoned module.

Reply:
Actually for DBIx::Placeholder::Named it would have been a lot more work that just updating a query string as the module is a wrapper around the larger DBI library methods which I wanted to avoid. So I just pulled out the small bit that I needed and then altered it some more to add ability to pass in a list of values instead of just single values.